deployment/state.go

```diff
@@ -0,0 +1,97 @@
+package deployment
+
+import (
+	"errors"
+	"reflect"
+)
+
+var (
+	ErrInvalidStateType = errors.New("invalid state type")
+	ErrStateNotFound    = errors.New("no state with this type can be found")
+
+	ErrStateManagerNotFound = errors.New("the environment did not include a state manager")
+)
+
+// StateKey is a carrier of type information, to allow a static type to be used as a key, conveniently, without
+// having to mess with the reflection package in client code. Its basic usage is:
+//
+//	key := StateKey[MyStruct]{}
+//
+// The type parameter information is carried on the inner field, and infrastructure code can obtain that type for use
+// in underlying code.
+type StateKey[T StateRepresentation] struct {
+	_ T // Expose the type to generics.
+}
+
+// StateManager is responsible for looking up and supplying state to the typesafe GetState wrapper function. Different
+// implementations will obtain the state (and populate the given struct) from different sources, whether in-memory,
+// from json files, via RPCs or any other mechanism. The simple InMemoryStateManager can be used for in-memory
+// Environments. More complex StateManager implementations can be found where those Environment variations are
+// implemented.
+type StateManager interface {
+	// getState returns a struct of the type given in the "key" parameter, populated with any state data obtained
+	// using that same key. If no such state exists, it returns ErrStateNotFound. If it finds any errors in finding
+	// and marshalling the data in question, it returns those errors.
+	getState(key reflect.Type) (StateRepresentation, error)
+}
+
+// InMemoryStateManager is a non-thread-safe StateManager implementation, which simply maintains an internal
+// map keyed by reflect.Type. The struct value of the map is required to be of the type represented by the key.
+//
+// This can be used in a test quite simply to setup state for use by the GetState function, outside the ChangeSet
+// under test, injected into the environment, and then the ChangeSet can access that state via the GetState function.
+type InMemoryStateManager struct {
+	// Maintains the map of type to state struct. Public for testing. This map is unavailable within a ChangeSet
+	State map[reflect.Type]StateRepresentation
```

Comment:
Is the idea to use this implementation for the current views? Will CLD build the state using the current renderers and construct the manager?